### ISSUE TITLE
I dunno what to call this. Mysterious object corruption/overwritten?

### ISSUE THREAD

**Issue by credomane** [nodejs/help#3961]
### Version

18.8.0

### Platform

Windows 10 Home

### What steps will reproduce the bug?

Look at/run the code in the linked gist. I don't understand bug well enough to reproduce it through explanation or even give a proper title, sorry. This bug has been driving me nuts for the past few days and until a couple hours ago I thought it was me somehow doing something goofy.

So this gist (at the bottom) is the minimal amount of code to cause this problem for me.
In the actual real data set the skyward and google constants are loaded from json files and contain a few thousand entries each. There are even a few hundred entries between the two items in the skyward set I've provided. In the full version of the main loop it splits the data into a few different groups: perfect, good, multi, lowconfidence, and nomatch. No group will have data shared with another group. Then all this data is exported to json files with the same names as the group name. Everything works great until 2+ skyward students have identical the same first and last name then all of a sudden the lowConfidence data is modified for those particular students with the data for the most recent one while all the others students in lowConfidence are left as they were. It doesn't make any sense. Especially after adding in the `console.log` calls.

The findmeInGoogle function is just a 'basic' comparison function and returns an object with some arbitrary confidence score and is working 100% of the time.

However the code after the call findmeInGoogle has an issue that I don't really know how to explain and it is driving me absolutely insane. In this minimal example the lowConfidence[0].google gets modified out of no where after the call to findmeInGoogle. The only thing I can think of is some bug in node itself. I've tried node 16.16.0 then updated to 16.17.0 and finally 18.8.0 and the problem persists.

For now I've had to work around this issue I've had to go the ugly route and `JSON.parse(JSON.stringify(match.results));` and assign that to the "new" entry to the lowconfidence array.

https://gist.github.com/credomane/1b61aef3c6b019ea236f27493d660978

### How often does it reproduce? Is there a required condition?

Every time if you use the linked gist code.

### What is the expected behavior?

```
[LOOPA] []
[LOOPB] []
[LOOPC] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
[LOOPA] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
[LOOPB] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
[LOOPC] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
[DONE] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
```

### What do you see instead?

```
[LOOPA] []
[LOOPB] []
[LOOPC] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
[LOOPA] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"name","__confidence":1,"__skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"}}]}]
[LOOPB] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"stuidnamegrademail","__confidence":10,"__skyward":{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","emailHome":"","gradyear":"2150","entityid":"200"}}]}]
[LOOPC] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"stuidnamegrademail","__confidence":10,"__skyward":{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","emailHome":"","gradyear":"2150","entityid":"200"}}]}]
[DONE] [{"success":true,"confidence":1,"skyward":{"stuid":"052364","active":false,"last":"JOHN","first":"DOE","email":"","emailHome":"","gradyear":"2005","entityid":"400"},"google":[{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","gradyear":"2150","org":"/Students","lastLogin":"2022-08-22T13:11:38.000Z","__method":"stuidnamegrademail","__confidence":10,"__skyward":{"stuid":"505910","active":true,"last":"JOHN","first":"DOE","email":"johnd50@stu.testdomain.com","emailHome":"","gradyear":"2150","entityid":"200"}}]}]
```

**Comment by credomane** [comment]
>people not realizing that objects passed to JavaScript functions are passed by reference. 

This was the very first thing I checked. While I'm not saying I'm not still screwing that up somewhere, I doubt it.

>Since this is extraordinarily unlikely to be a bug in Node.js--and even if it is a bug in the runtime, which I very much doubt, that bug would almost certainly be in V8 and not in Node.js itself--I'm going to move this issue to the help repo.

So that would mean the issue should also be present in chrome then correct? and potentionally NOT in firefox? I'll give them both a shot and see what happens since my gist code should be free of anything node specific.

**Comment by Trott** [comment]
> So that would mean the issue should also be present in chrome then correct? and potentionally NOT in firefox? I'll give them both a shot and see what happens since my gist code should be free of anything node specific.

If you can't reproduce it in Chrome but can reproduce it in Node.js, that would be an indication the problem is in Node.js.

If it is reproduced in Chrome and Firefox, I'd take that to mean that it is extremely likely that the problem is in your code somewhere and not in V8 or Node.js. 

Also, Node.js comes with a command-line step debugger as well as a connect-your-Chrome-browser-for-debugging-purposes inspector feature. Those might be useful here. Add a watch to the value that you are surprised by and see when it changes.

**Comment by credomane** [comment]
>Also, Node.js comes with a command-line step debugger as well 

...I really need to learn how to do that. Guess now's the time because I really want to know what the deal is.

**Comment by Trott** [comment]
The command-line debugger is very minimal, but is probably enough for your gist. You can access it by running your script with `node inspect scriptname.js` instead of `node scriptname.js`. The docs are at https://nodejs.org/dist/latest-v18.x/docs/api/debugger.html.

**Comment by credomane** [comment]
Took me way too long to figure out the debug commands because the step in/out of functions would "break" my watchers. Once I got that worked out...Yes, it ended up being an accidental modification to a sub-value on a reference object. The modification was intentional but I meant to modify the clone...the clone that I never made. I'm so pissed at myself right now. Soon as I saw the line come up in the debugger even before it was executed I knew it was my problem. Can't believe I missed it all this time in vscode. So many hours needlessly spent in frustration.

Sorry for wasting your time.
